[pytree] test aligned API signature for C++ and Python pytree

XuehaiPan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #114858
* #113258
* #113257
* #113256
* #113255
* __->__ #112485

Add tests to ensure the C++ and Python pytree provide the same APIs with identical signatures.

cc @zou3519 @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4

XuehaiPan committed:
[pytree] test aligned API signature for C++ and Python pytree

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519 avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi suo ydwu4

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519 avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi suo ydwu4

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519 avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi suo ydwu4

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519 avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi suo ydwu4

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519 avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi suo ydwu4

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519 avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi suo ydwu4

[ghstack-poisoned]

PaliC commented:
@pytorchbot revert -m "in the bottom diff in the stack changing _register_pytree_node's signature is bc breaking, please revert the signature and reland" -c ghfirst

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519 avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi suo ydwu4

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519 avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi suo ydwu4

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519 avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi suo ydwu4

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519 avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi suo ydwu4

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519 avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi suo ydwu4

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519 avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi suo ydwu4

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519 avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi suo ydwu4

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519 avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi suo ydwu4

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519 avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi suo ydwu4

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"




cc zou3519 avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi suo ydwu4

[ghstack-poisoned]

XuehaiPan commented:
I reordered the commits in the gh stack.

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"


Add tests to ensure the C++ and Python pytree provide the same APIs with identical signatures.

cc zou3519 avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi suo ydwu4

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"


Add tests to ensure the C++ and Python pytree provide the same APIs with identical signatures.

cc zou3519 avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi suo ydwu4

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"


Add tests to ensure the C++ and Python pytree provide the same APIs with identical signatures.

cc zou3519 avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi suo ydwu4

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"


Add tests to ensure the C++ and Python pytree provide the same APIs with identical signatures.

cc zou3519 avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi suo ydwu4

[ghstack-poisoned]

zou3519 commented on test/test_pytree.py:
```diff
@@ -26,6 +28,98 @@ def __init__(self, x, y):
+                cxx_signature = inspect.signature(cxx_api)
+                py_signature = inspect.signature(py_api)
+
+                # The C++ pytree APIs provide more features than the Python APIs.
+                # The Python APIs are a subset of the C++ APIs.
+                # Check the signature of the Python API is a subset of the C++ API.
+                cxx_param_names = list(cxx_signature.parameters)
+                py_param_names = list(py_signature.parameters)
+                self.assertTrue(
+                    set(cxx_param_names).issuperset(py_param_names),
+                    msg=(
+                        f"C++ parameter(s) ({cxx_param_names}) "
+                        f"not in Python parameter(s) ({py_param_names})"
+                    ),
+                )
```
What are the examples? Ideally we don't diverge at all in the APIs we provide, otherwise there's no swapping between python pytree and C++ pytree

XuehaiPan replied:
The C++ pytree provides extra `None` type handling and namespaced registry support. There are three extra features in the C++ pytree function signature:

1. Extra `namespace` (defaults to `"torch"`) keyword argument in all tree operation functions and `register_pytree_node`.
2. Extra `none_is_leaf` (defaults to `True`) keyword argument in all tree operation functions.
3. C++ pytree supports multi-tree-map, while Python pytree only supports mapping over a single pytree.

- pytorch/pytorch#91323

```python
# Extra argument `namespace`
def cxx_register_pytree_node(cls, flatten_fn, unflatten_fn, ...other args..., namespace="torch"): ...
def python_register_pytree_node(cls, flatten_fn, unflatten_fn, ...other args...): ...

# Extra argument `none_is_leaf` and `namespace`
def cxx_tree_flatten(tree, *, none_is_leaf=True, namespace="torch"): ...
def python_tree_flatten(tree): ...

# Extra argument `none_is_leaf` and `namespace` and multi-tree-map support
def cxx_tree_map(func, tree, *rests, none_is_leaf=True, namespace="torch"): ...
def python_tree_map(func, tree): ...
```

For 1 and 2, I can remove `none_is_leaf` and `namespace` arguments and hardcode them to their default values.

zou3519 replied:
Thank you

zou3519 reviewed (commented):
PR looks mostly good to me, but I'm confused about why some of the C++ pytree APIs are a superset of the python pytree APIs

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"


Add tests to ensure the C++ and Python pytree provide the same APIs with identical signatures.

cc zou3519 avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi suo ydwu4

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] test aligned API signature for C++ and Python pytree"


Add tests to ensure the C++ and Python pytree provide the same APIs with identical signatures.

cc zou3519 avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi suo ydwu4

[ghstack-poisoned]

XuehaiPan commented:
@pytorchbot merge
